### PR TITLE
Add t0 scripts

### PR DESCRIPTION
**Pull request by Muennighoff**
Notes:

RE: Learning Rate
T0 & FLAN use Adafactor which automatically adjusts the step size:
[Finally, while the learning rate in Adam denotes a target absolute step size, we follow the intuition that relative change in the parameters is more relevant, so we propose scaling the size of the updates relative to the scale of the parameters themselves](https://arxiv.org/pdf/1804.04235.pdf). Due to this scaling Adafactor may more resistent to higher learning rates and the step size adjusts automatically, so scheduling may be less needed (I.e. if you have weight decay with Adafactor, step size will automatically decay because parameters decay). For now I'm keeping a constant conservative LR of `1e-5`, but we may want to instead go higher and add warmup + scheduling. Thoughts?

### CODE REVIEW

**Comment by thomasw21** [comment]
> T0 leaves some HPs unspecified like Warmup, Weight Decay; Let's discuss them here?

No warmup as you have constant learning rate.
No weight decay (will double check that one)

> Currently, it would use 5% of the training set for validation.

If you can actually use the validation data from T0 then I'd say this is better.

**Comment by Muennighoff** [comment]
> If you can actually use the validation data from T0 then I'd say this is better.

For that either
a) Add a new arg like `args.data_path` that calls build_train_valid_test_datasets again with 100% valid split
b) Concat train & valid sets, make on indexed dataset & the ratio such that they are separated again
c) Use `args.valid_weighted_split_paths` & `build_dataset_group`, which doesn't work yet for MTF

I think a) or c) is best - Wdyt?

**Comment by thomasw21** [comment]
Probably need to use this: it's already implemented as an API https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/c5b88fb92d4417f77d729c95ce95e3a740b47065/megatron/arguments.py#L822-L840, I'll update the T0 branch to have that feature.

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 1]
```diff
@@ -80,7 +79,6 @@ OPTIMIZER_ARGS=" \
     --adam-eps 1e-8 \
     --lr 1e-3 \
     --lr-decay-style constant \
-    --lr-warmup-samples $LR_WARMUP_SAMPLES \
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm]: We used `Adafactor` ... so technically I don't know what parameters matter (typically we used a `decay` argument, which I don't know how it translates to `Adam` optimizer)

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 2]
```diff
@@ -80,7 +79,6 @@ OPTIMIZER_ARGS=" \
...
     --clip-grad 1.0 \
     --weight-decay 1e-1 \
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:83]: This is set to 0

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 3]
```diff
@@ -0,0 +1,198 @@
+#!/bin/bash
+#SBATCH --job-name=tr11f-6B3-ml-t0
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:2]: We should rename this to tr13 or something.

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+set -x -e
+
+source $six_ALL_CCFRWORK/start-tr11f-6B3-ml-t0 # TODO
+echo "START TIME: $(date)"
+
+variant=main
+
+DATA_OUTPUT_PATH=$six_ALL_CCFRSCRATCH/checkpoints/tr11f-6B3-ml-t0
+CHECKPOINT_PATH=$DATA_OUTPUT_PATH/checkpoints/$variant
+REPO_PATH=$DATA_OUTPUT_PATH/tr11f-6B3-ml-t0-logs
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:25]: Ditto: `tr13f-6b3-ml-t0-logs` and we specify in the ReadMe what tr13 stands for (MTF)

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+TENSORBOARD_PATH=$REPO_PATH/tensorboard/$variant
+LOGS_PATH=$REPO_PATH/logs/$variant
+mkdir -p $LOGS_PATH
+
+MEGATRON_DEEPSPEED_REPO=$six_ALL_CCFRWORK/code/tr11f-6B3-ml-t0/Megatron-DeepSpeed # TODO
+cd $MEGATRON_DEEPSPEED_REPO
+
+BIGSCIENCE_REPO=$six_ALL_CCFRWORK/code/bigscience
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:33]:
```suggestion
BIGSCIENCE_REPO=$six_ALL_CCFRWORK/code/tr11f-6B3-ml-t0/bigscience
```

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+TOKENIZER_NAME_OR_PATH=bigscience-catalogue-data-dev/byte-level-bpe-tokenizer-no-norm-250k-whitespace-and-eos-regex-alpha-v3-dedup-lines-articles
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:34]:
```suggestion
TOKENIZER_NAME_OR_PATH=bigscience/tokenizer
```
We should make the switch.

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+GLOBAL_BATCH_SIZE=512
+
+NLAYERS=30
+NHIDDEN=4096
+NHEADS=32
+SEQ_LEN=2048
+
+SAVE_INTERVAL=500
+
+TRAIN_SAMPLES=6_400_000  # 13e9 / 2048
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:71]: Let's make that a multiple of 512 no?

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+# We use 2048 total tokens and 512 batch size = 2**20
+MICRO_BATCH_SIZE=1
+GLOBAL_BATCH_SIZE=512
+
+NLAYERS=30
+NHIDDEN=4096
+NHEADS=32
+SEQ_LEN=2048
+
+SAVE_INTERVAL=500
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:69]: We should probably have a bigger value no? something like 1000-1500 (so we can have only 13 checkpoints?)

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    --pad-vocab-size-to 250880 \
+    $OPTIMIZER_ARGS \
+    $EXIT_OPTS \
+    "
+
+OUTPUT_ARGS=" \
+    --log-interval 1 \
+    --save-interval $SAVE_INTERVAL \
+    --eval-interval 1000 \
+    --eval-iters 1 \
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:118]: Maybe a bit more no? 10?

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    --tensorboard-dir $TENSORBOARD_PATH \
+    --tensorboard-queue-size 5 \
+    --log-timers-to-tensorboard \
+    --log-batch-size-to-tensorboard \
+    --log-validation-ppl-to-tensorboard \
+    "
+
+ZERO_STAGE=1 # important: bf16 must use z0! it implements its own zero stage 1 equivalent
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:126]:
```suggestion
ZERO_STAGE=1
```

File: train/t0/tr11f-6B3-ml-t0.slurm [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+export CMD=" \
+    `pwd`/finetune_t0_non_causal_decoder.py \
+    --tensor-model-parallel-size $TP_SIZE \
+    --pipeline-model-parallel-size $PP_SIZE \
+    $GPT_ARGS \
+    $OUTPUT_ARGS \
+    --save $CHECKPOINT_PATH \
+    --load $CHECKPOINT_PATH \
+    --data-path $DATA_PATH \
+    --split 95,5,0 \
```
Comment (thomasw21) [inline train/t0/tr11f-6B3-ml-t0.slurm:181]: Change to other mechanism.

File: train/t0/tr11f-6B3-ml-t0.slurm (hunk 9 above)
Comment (Muennighoff) [reply train/t0/tr11f-6B3-ml-t0.slurm]: Set it to 10

File: train/tr13-t0/t0_test.slurm [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+#SBATCH --account=six@v100
+
+set -x -e
+
+source $six_ALL_CCFRWORK/start-muennighofflmeval
+echo "START TIME: $(date)"
+
+variant=main
+
+DATA_OUTPUT_PATH=$six_ALL_CCFRSCRATCH/checkpoints/tr11e-350M-ml
```
Comment (thomasw21) [inline train/tr13-t0/t0_test.slurm:20]:
```suggestion
DATA_OUTPUT_PATH=$six_ALL_CCFRSCRATCH/checkpoints/tr13-test-ml
```

File: train/tr13-t0/tr13f-6B3-ml-t0.slurm [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    --nproc_per_node $GPUS_PER_NODE \
+    --nnodes $NNODES \
+    --rdzv_endpoint $MASTER_ADDR:$MASTER_PORT \
+    --rdzv_backend c10d \
+    --max_restarts 0 \
+    --tee 3 \
+    "
+
+export CMD=" \
+    `pwd`/finetune_t0_non_causal_decoder.py \
```
Comment (thomasw21) [inline train/tr13-t0/tr13f-6B3-ml-t0.slurm:183]:
```suggestion
    `pwd`/finetune_t0_causal_decoder.py \
```
Right now all the script use `is_causal=True` so we should rename this in Meg-DS PR.

**Review by thomasw21 — APPROVED** [review]
Looked mostly at 6B3 config. Seems alright. Thanks!

File: train/tr13-t0/tr13f-6B3-ml-t0.slurm (hunk 13 above)
Comment (Muennighoff) [reply train/tr13-t0/tr13f-6B3-ml-t0.slurm:183]: Added an arg [here](https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/309); Lets merge that PR first before we merge here

**Comment by Muennighoff** [comment]
Already merged via other PR
